File: Easy/20. Valid Parentheses/ValidParentheses.py
'''
LeetCode Easy: 20. Valid Parentheses
https://leetcode.com/problems/valid-parentheses/

Check ValidParentheses.cpp for brief solution explanation.
'''
class Solution:
    def isValid(self, s: str) -> bool:
        N = len(s)
        m = {'{': 1, '}': 1, '(': 2, ')': 2, '[': 3, ']': 3}
        open_brackets = []
        for i in range(N):
            cur = m[s[i]]
            if s[i] in "{([":
                open_brackets.append(cur)
            else:
                if len(open_brackets)<=0 or open_brackets[-1]!=cur:
                    return False
                else:
                    open_brackets.pop()
        return len(open_brackets)==0

File: Easy/20. Valid Parentheses/ValidParentheses.java
/*
LeetCode Easy: 20. Valid Parentheses
https://leetcode.com/problems/valid-parentheses/

Check ValidParentheses.cpp for brief solution explanation.
*/
class Solution {
    public boolean isValid(String s) {
        int N = s.length();
        HashMap<Character, Integer>m = new HashMap<Character, Integer>();
        m.put('{', 1); m.put('}', 1);
        m.put('(', 2); m.put(')', 2);
        m.put('[', 3); m.put(']', 3);
        Stack<Integer>open = new Stack<Integer>();
        for(int i = 0;i<N;i++){
            int cur = m.get(s.charAt(i));
            if(s.charAt(i)=='{'||s.charAt(i)=='('||s.charAt(i)=='['){
                open.add(cur);
            }else{
                if(open.size()<=0||open.peek()!=cur){
                    return false;
                }else{
                    open.pop();
                }
            }
        }
        return open.size()==0;
    }
}